0099-Replacing bash as the default shell.md

```diff
@@ -0,0 +1,120 @@
+
+[reference](https://github.com/oilshell/oil/wiki/Migration-Guide)
+
+Some more background context. Originally this is coming from this [PR](https://github.com/NixOS/nixpkgs/pull/105233) by zimbatm. After seeing this, I thought I would implement the changes needed to make the switch and generate a discussion in a PR. Somebody brought to my attention that this change is significant enought that it needs to go through an RFC.
+
+## Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+Some extracts from the post [why use oil](https://www.oilshell.org/why.html)
+
+- better error handling. One of the confusing things about bash is how to handle errors in scripts. `err_exit` was a try at improving that but it has many pitfalls. Oil aims to have more straightforward error handling. Any error will exit, if you do not want that behavior there is the `try my_function` form.
+
+`mkdir /tmp/dest && cp foo /tmp/dest` becomes simply
+
+```Shell
+mkdir /tmp/dest
+cp foo /tmp/dest
+```
+
+- Introduce functions that need to explicitely define their parameters
+
+```Shell
+proc f(first, @rest) { # @ means "the rest of the arguments"
+  write -- $first
+  write -- @rest # @ means "splice this array"
+}
+```
+
+instead of the traditional
+
+```Shell
+f() {
+  echo $1
+  shift 1
+  echo "$@"
+}
+```
+
+those functions are called procs and their variables don't mess with their outer scope.
+
+- Oil has proper associative arrays (dictionaries or hashmaps) that don't have the problems that bash's have.
+
+- No more quotes to prevent splitting
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Currently oilshell has a bus factor of 1. While Andrew has been very motivated and responsive thus far, this could change in the future.
+- Oil shell is pre 1.0 so the syntax might change.
+- Parts of oil are rewritten in c++, while this could be good for performance, new user might discover bugs that are really hard to debug.
+- Oil has had little usage so far, there might be bugs that will just be discovered by having a large user base.
+- Using osh, could be straightforward, but using oil might take more work as basically the existing bash code need to rewritten with `strict_errexit`.
+- The oil documentation is lacking at the moment.
+- Oil has really lofty goals, which I think it's good, but some potential solutions are not documented or not implemented. One example is eggex which are meant to be a replacement for reggexes.
+
+## Alternatives
```

Comment:
I really like the idea of the flag. I would even want to have it for specific packages for people to use their favorite shell. The only thing would be that if people can start using their own shell for a package and they use specific features of that shell, then we will never be able to remove that flag again (I don't necessarily think it's a bad idea). Perhaps you were thinking of a flag that isn't changeable by individual developpers ?

Reply:
I would hesitate about "for specific packages for people to use their favorite shell". This puts long-term maintenance burden on stdenv and hooks to work with as many shells as possible, it also means that we can't take advantage of the convenience and safety features of oil. I don't think that is worth it. I see it purely as a migration tool between `bash` and `osh`. It could be set per-package so that the migration looks something like this.

1. Set `shell = "oil"` for a small number of packages after updating them and stdenv to work.
2. Expand this list of packages until we believe that just about everything in stdenv and popular hooks are working on oil.
3. Flip the default shell and add `shell = "bash"` to any packages that this breaks.
4. Burn down the list of `shell = "bash"` packages to zero.
5. Remove the `shell` flag from stdenv. ("inline" oil as the default shell)
6. We can now take advantage of oil features!

The ensures that each step is small and safe and enforces forward progress through the migration (accidental regressions on either shell can't happen as some builds will fail).

Again, I recommend this only as a migration tool. I think the long-term maintenance of multiple shells is likely more effort than it is worth.